src/commands/shell.rs

```diff
@@ -73,20 +93,135 @@ pub async fn command(args: Args, _json: bool) -> Result<()> {
         eprintln!("No service linked, skipping service variables");
     }
 
+    // disable dead code warning for windows_shell_detection
+    #[allow(dead_code)]
+    enum WindowsShell {
+        Cmd,
+        Powershell,
+        Powershell7,
+    }
+
+    /// https://gist.github.com/mattn/253013/d47b90159cf8ffa4d92448614b748aa1d235ebe4
+    /// Only matches cmd, powershell or pwsh for safety
+    /// defaults to cmd if no parent process is found
+    #[cfg(target_os = "windows")]
+    async fn windows_shell_detection() -> Option<WindowsShell> {
+        let (_, ppname) = get_parent_process_info()
+            .context("Failed to get parent process info")
+            .unwrap_or_else(|_| (0, "".to_string()));
+
+        if ppname.contains("pwsh") {
+            Some(WindowsShell::Powershell7)
+        } else if ppname.contains("powershell") {
+            Some(WindowsShell::Powershell)
+        } else {
+            Some(WindowsShell::Cmd)
+        }
+    }
+
+    #[cfg(not(target_os = "windows"))]
+    async fn windows_shell_detection() -> Option<WindowsShell> {
+        None
+    }
+
     let shell = std::env::var("SHELL").unwrap_or(match std::env::consts::OS {
-        "windows" => "cmd".to_string(),
+        "windows" => match windows_shell_detection().await {
+            Some(WindowsShell::Powershell) => "powershell".to_string(),
+            Some(WindowsShell::Cmd) => "cmd".to_string(),
+            Some(WindowsShell::Powershell7) => "pwsh".to_string(),
+            None => "cmd".to_string(),
+        },
         _ => "sh".to_string(),
     });
 
-    println!("Entering subshell with Railway variables available. Type 'exit' to exit.");
+    let shell_options = match shell.as_str() {
+        "powershell" => vec!["/nologo"],
+        "pwsh" => vec!["/nologo"],
+        "cmd" => vec!["/k"],
+        _ => vec![],
+    };
+
+    println!("Entering subshell with Railway variables available. Type 'exit' to exit.\n");
 
     tokio::process::Command::new(shell)
+        .args(shell_options)
         .envs(all_variables)
         .spawn()
         .context("Failed to spawn command")?
         .wait()
         .await
         .context("Failed to wait for command")?;
+
     println!("Exited subshell, Railway variables no longer available.");
     Ok(())
 }
+
+/// get the parent process info, translated from
+// https://gist.github.com/mattn/253013/d47b90159cf8ffa4d92448614b748aa1d235ebe4
+#[cfg(target_os = "windows")]
+fn get_parent_process_info() -> Option<(DWORD, String)> {
```

Comment:
This whole function should probably be marked as unsafe.

Reply:
Marked as unsafe